gui/src/app/SamplerOutputView/stan_stats/stan_stats.ts

```diff
@@ -0,0 +1,318 @@
+/*
+Translated from
+https://github.com/stan-dev/stan/blob/develop/src/stan/analyze/mcmc/compute_effective_sample_size.hpp
+and
+https://github.com/stan-dev/stan/blob/develop/src/stan/analyze/mcmc/autocovariance.hpp
+and
+https://github.com/stan-dev/stan/blob/develop/src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp
+*/
+
+import { transform as inPlaceFftTransform, inverseTransform as inPlaceInverseFftTransform } from "./fft";
+
+/**
+ * Computes the effective sample size (ESS) for the specified
+ * parameter across all kept samples.  The value returned is the
+ * minimum of ESS and the number_total_draws *
+ * log10(number_total_draws).
+ *
+ * See more details in Stan reference manual section "Effective
+ * Sample Size". http://mc-stan.org/users/documentation
+ *
+ * Current implementation assumes draws are stored in contiguous
+ * blocks of memory.  Chains are trimmed from the back to match the
+ * length of the shortest chain.  Note that the effective sample size
+ * can not be estimated with less than four draws.
+ *
+ * draws: arrays of draws for each chain
+ * returns: effective sample size for the specified parameter
+ */
+export function compute_effective_sample_size(draws: number[][]): number {
+    const num_chains = draws.length;
+
+    // use the minimum number of draws across all chains
+    let num_draws = draws[0].length;
+    for (let chain = 1; chain < num_chains; ++chain) {
+        num_draws = Math.min(num_draws, draws[chain].length);
+    }
+
+    if (num_draws < 4) {
+        // we don't have enough draws to compute ESS
+        return NaN;
+    }
+
+    // check if chains are constant; all equal to first draw's value
+    let are_all_const = false;
+    const init_draw = new Array(num_chains).fill(0);
+    for (let chain_idx = 0; chain_idx < num_chains; chain_idx++) {
+        const draw = draws[chain_idx];
+        for (let n = 0; n < num_draws; n++) {
+            if (!isFinite(draw[n])) {
+                // we can't compute ESS if there are non-finite values
+                return NaN;
+            }
+        }
+
+        init_draw[chain_idx] = draw[0];
+
+        const precision = 1e-12;
+        if (draw.every(d => Math.abs(d - draw[0]) < precision)) {
+            are_all_const = true;
+        }
+    }
+
+    if (are_all_const) {
+        // If all chains are constant then return NaN
+        // if they all equal the same constant value
+        const precision = 1e-12;
+        if (init_draw.every(d => Math.abs(d - init_draw[0]) < precision)) {
+            return NaN;
+        }
+    }
```

Comment:
Similarly, want to pull this into a helper? I realize this is also a suggestion I could be making to the Stan implementation...

Reply:
I think it's best to try to match the C++ implementation as closely as possible, even with the imperfections. If the goal is to make it cleaner, there's a lot that could be done... but I think the goal here is to reproduce exactly.